test/libdevcrypto/LibSnark.cpp

```diff
@@ -0,0 +1,470 @@
+
+	ss << fh.rdbuf();
+	fh.close();
+
+	ss.rdbuf()->pubseekpos(0, std::ios_base::in);
+
+	libsnark::r1cs_ppzksnark_verification_key<libsnark::alt_bn128_pp> verificationKey;
+	ss >> verificationKey;
+
+//	testProof(verificationKey);
+//	return;
+
+	unsigned icLength = verificationKey.encoded_IC_query.rest.indices.size() + 1;
+
+
+	cout << "\tfunction verifyingKey() internal returns (VerifyingKey vk) {" << endl;
+	cout << "\t\tvk.A = " << outputPointG2Affine(verificationKey.alphaA_g2) << ";" << endl;
+	cout << "\t\tvk.B = " << outputPointG1Affine(verificationKey.alphaB_g1) << ";" << endl;
+	cout << "\t\tvk.C = " << outputPointG2Affine(verificationKey.alphaC_g2) << ";" << endl;
+	cout << "\t\tvk.gamma = " << outputPointG2Affine(verificationKey.gamma_g2) << ";" << endl;
+	cout << "\t\tvk.gammaBeta1 = " << outputPointG1Affine(verificationKey.gamma_beta_g1) << ";" << endl;
+	cout << "\t\tvk.gammaBeta2 = " << outputPointG2Affine(verificationKey.gamma_beta_g2) << ";" << endl;
+	cout << "\t\tvk.Z = " << outputPointG2Affine(verificationKey.rC_Z_g2) << ";" << endl;
+	cout << "\t\tvk.IC = new Pairing.G1Point[](" << icLength << ");" << endl;
+	cout << "\t\tvk.IC[0] = " << outputPointG1Affine(verificationKey.encoded_IC_query.first) << ";" << endl;
+	for (size_t i = 1; i < icLength; ++i)
+	{
+		auto vkICi = outputPointG1Affine(verificationKey.encoded_IC_query.rest.values[i - 1]);
+		cout << "\t\tvk.IC[" << i << "] = " << vkICi << ";" << endl;
+	}
+	cout << "\t\t}" << endl;
+	cout << "\tfunction verify(uint[] input, Proof proof) internal returns (bool) {" << endl;
+	cout << "\t\tVerifyingKey memory vk = verifyingKey();" << endl;
+	cout << "\t\tif (input.length + 1 != vk.IC.length) throw;" << endl;
+	cout << "\t\t// Compute the linear combination vk_x" << endl;
+	cout << "\t\tPairing.G1Point memory vk_x = vk.IC[0];" << endl;
+	for (size_t i = 0; i < verificationKey.encoded_IC_query.rest.indices.size(); ++i)
+		cout << "\t\tvk_x = Pairing.add(vk_x, Pairing.mul(vk.IC[" << (i + 1) << "], input[" << i << "]));" << endl;
+	cout << "\t\tif (!Pairing.pairingProd2(proof.A, vk.A, Pairing.negate(proof.A_p), Pairing.P2())) return false;" << endl;
+	cout << "\t\tif (!Pairing.pairingProd2(vk.B, proof.B, Pairing.negate(proof.B_p), Pairing.P2())) return false;" << endl;
+	cout << "\t\tif (!Pairing.pairingProd2(proof.C, vk.C, Pairing.negate(proof.C_p), Pairing.P2())) return false;" << endl;
+	cout << "\t\tif (!Pairing.pairingProd3(proof.K, vk.gamma, Pairing.negate(Pairing.add(vk_x, Pairing.add(proof.A, proof.C))), vk.gammaBeta2, Pairing.negate(vk.gammaBeta1), proof.B)) return false;" << endl;
+	cout << "\t\tif (!Pairing.pairingProd3(" << endl;
+	cout << "\t\t\t\tPairing.add(vk_x, proof.A), proof.B," << endl;
+	cout << "\t\t\t\tPairing.negate(proof.H), vk.Z," << endl;
+	cout << "\t\t\t\tPairing.negate(proof.C), Pairing.P2()" << endl;
+	cout << "\t\t)) return false;" << endl;
+	cout << "\t\treturn true;" << endl;
+	cout << "\t}" << endl;
+}
```

Comment:
Why is this commented out?  Does it make sense to turn this into a C++ template?